app/scripts/controllers/metamask-notifications/metamask-notifications.ts

```diff
@@ -0,0 +1,833 @@
+        'AuthenticationController:getBearerToken',
+      );
+    },
+    isSignedIn: () => {
+      return this.messagingSystem.call('AuthenticationController:isSignedIn');
+    },
+  };
+
+  #storage = {
+    getStorageKey: () => {
+      return this.messagingSystem.call('UserStorageController:getStorageKey');
+    },
+    getNotificationStorage: async () => {
+      return await this.messagingSystem.call(
+        'UserStorageController:performGetStorage',
+        'notification_settings',
+      );
+    },
+    setNotificationStorage: async (state: string) => {
+      return await this.messagingSystem.call(
+        'UserStorageController:performSetStorage',
+        'notification_settings',
+        state,
+      );
+    },
+  };
+
+  #pushNotifications = {
+    enablePushNotifications: async (UUIDs: string[]) => {
+      return await this.messagingSystem.call(
+        'PushPlatformNotificationsController:enablePushNotifications',
+        UUIDs,
+      );
+    },
+    disablePushNotifications: async (UUIDs: string[]) => {
+      return await this.messagingSystem.call(
+        'PushPlatformNotificationsController:disablePushNotifications',
+        UUIDs,
+      );
+    },
+    updatePushNotifications: async (UUIDs: string[]) => {
+      return await this.messagingSystem.call(
+        'PushPlatformNotificationsController:updateTriggerPushNotifications',
+        UUIDs,
+      );
+    },
+  };
+
+  #prevAccountsSet = new Set<string>();
+
+  #accounts = {
+    /**
+     * Used to get list of addresses from keyring (wallet addresses)
+     *
+     * @returns addresses removed, added, and latest list of addresses
+     */
+    listAccounts: async () => {
+      const nonChecksumAccounts = await this.messagingSystem.call(
+        'KeyringController:getAccounts',
+      );
+      const accounts = nonChecksumAccounts.map((a) => toChecksumHexAddress(a));
+      const currentAccountsSet = new Set(accounts);
+
+      const accountsAdded = accounts.filter(
+        (a) => !this.#prevAccountsSet.has(a),
+      );
+
+      const accountsRemoved = [...this.#prevAccountsSet.values()].filter(
+        (a) => !currentAccountsSet.has(a),
+      );
+
+      this.#prevAccountsSet = new Set(accounts);
+      return {
+        accountsAdded,
+        accountsRemoved,
+        accounts,
+      };
+    },
+
+    /**
+     * Initializes the cache/previous list. This is handy so we have an accurate in-mem state of the previous list of accounts.
+     *
+     * @returns result from list accounts
+     */
+    initialize: () => {
+      return this.#accounts.listAccounts();
+    },
+
+    /**
+     * Subscription to any state change in the keyring controller (aka wallet accounts).
+     * We can call the `listAccounts` defined above to find out about any accounts added, removed
+     * And call effects to subscribe/unsubscribe to notifications.
+     */
+    subscribe: () => {
+      this.messagingSystem.subscribe(
+        'KeyringController:stateChange',
```

Comment:
Sweet - looks around and I think the keyring controller events/actions are better suited for our use case. Tested & seems to work well!

- It targets wallet specific addresses that we want to track (not snap or other accounts)

- We can use the state change event to then check when new addresses are added or removed (see logic below)
  - there is a account removed event, but no account added event. But the below logic works nicely!

Reply:
yeah, that makes sense! good find